### PR TITLE
Remove hardlinking in Shipper code.

### PR DESCRIPTION
**Pull request by pstibrany**
#### What this PR does

This PR modifies Shipper to avoid doing local copy of a block (using hardlinks) before uploading the block. Shipper was also updating its own copy of meta.json file. This has been replaced with new `tsdb.UploadBlock` method which uploads `meta.json` file from in-memory structure instead of using meta.json file from disk.

Why cannot we update local `meta.json` file for block loaded by ingester? TSDB library tries to reload the blocks at some points. When updating local `meta.json` file, shipper deletes old version first and then uses Replace call from temp file to `meta.json`. If TSDB library tried to reload the blocks while `meta.json` is missing, it would unload the block from memory instead (and possibly load it again later when finding the block again).

This PR copies `block.Upload` function from Thanos into `tsdb.UploadBlock`. This version of the function takes metadata to upload as a parameter.

#### Which issue(s) this PR fixes or relates to

Follow up to https://github.com/grafana/mimir/pull/1957#discussion_r884932268

#### Checklist

- [x] Tests updated
- [na] Documentation added
- [na] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by replay** [comment]
> This PR modifies Shipper to avoid doing local copy of a block (using hardlinks) before uploading the block. Shipper was also updating its own copy of meta.json file. This has been replaced with

This sentence seems unfinished

File: pkg/ingester/shipper.go [hunk 1]
```diff
@@ -185,42 +187,22 @@ func (s *Shipper) Sync(ctx context.Context) (uploaded int, err error) {
 	return uploaded, nil
 }
 
-// sync uploads the block if not exists in remote storage.
-// TODO(khyatisoneji): Double check if block does not have deletion-mark.json for some reason, otherwise log it or return error.
+// update local meta.json file with Thanos-specific details, and upload the block
+// to blocks storage.
 func (s *Shipper) upload(ctx context.Context, meta *metadata.Meta) error {
```
Comment (replay) [inline pkg/ingester/shipper.go:192]: according to "effective go" the method comments should start with the name of the method:

```
Doc comments work best as complete sentences, which allow a wide variety of automated presentations. The first sentence should be a one-sentence summary that starts with the name being declared.
```

[ref](https://go.dev/doc/effective_go#commentary)

File: pkg/ingester/shipper.go [hunk 2]
```diff
@@ -185,42 +187,22 @@ func (s *Shipper) Sync(ctx context.Context) (uploaded int, err error) {
 	return uploaded, nil
 }
 
-// sync uploads the block if not exists in remote storage.
-// TODO(khyatisoneji): Double check if block does not have deletion-mark.json for some reason, otherwise log it or return error.
+// update local meta.json file with Thanos-specific details, and upload the block
+// to blocks storage.
```
Comment (replay) [inline pkg/ingester/shipper.go:191]: according to "effective go" the method comments should start with the name of the method:

```
Doc comments work best as complete sentences, which allow a wide variety of automated presentations. The first sentence should be a one-sentence summary that starts with the name being declared.
```

[ref](https://go.dev/doc/effective_go#commentary)

so something like:

```suggestion
// upload updates the local meta.json file with Thanos-specific details, and uploads
// the block to blocks storage.
```

File: pkg/ingester/shipper.go [hunk 3]
```diff
@@ -185,42 +187,22 @@ func (s *Shipper) Sync(ctx context.Context) (uploaded int, err error) {
...
-		}
-	}()
+	blockDir := filepath.Join(s.dir, meta.ULID.String())
 
-	dir := filepath.Join(s.dir, meta.ULID.String())
-	if err := hardlinkBlock(dir, updir); err != nil {
-		return errors.Wrap(err, "hard link block")
-	}
-	// Attach current labels and write a new meta file with Thanos extensions.
+	// Attach current labels and write. Don't write the file to disk, but upload it to the bucket.
```
Comment (replay) [inline pkg/ingester/shipper.go:197]: `and write`

I'm not sure I understand what the "write" is referring to

File: pkg/storage/tsdb/upload.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+	// Note that entry for meta.json file will be incorrect and will reflect local file,
+	// not updated Meta struct.
+	meta.Thanos.Files, err = gatherFileStats(blockDir)
+	if err != nil {
+		return errors.Wrap(err, "gather meta file stats")
+	}
+
+	metaEncoded := strings.Builder{}
+	if err := meta.Write(&metaEncoded); err != nil {
```
Comment (replay) [inline pkg/storage/tsdb/upload.go:60]:
```suggestion
	if err = meta.Write(&metaEncoded); err != nil {
```

File: pkg/storage/tsdb/upload.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+		return errors.Wrap(err, "encode meta file")
+	}
+
+	if err := objstore.UploadDir(ctx, logger, bkt, filepath.Join(blockDir, block.ChunksDirname), path.Join(id.String(), block.ChunksDirname)); err != nil {
```
Comment (replay) [inline pkg/storage/tsdb/upload.go:64]:
```suggestion
	if err = objstore.UploadDir(ctx, logger, bkt, filepath.Join(blockDir, block.ChunksDirname), path.Join(id.String(), block.ChunksDirname)); err != nil {
```

File: pkg/storage/tsdb/upload.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+		return cleanUp(logger, bkt, id, errors.Wrap(err, "upload chunks"))
+	}
+
+	if err := objstore.UploadFile(ctx, logger, bkt, filepath.Join(blockDir, block.IndexFilename), path.Join(id.String(), block.IndexFilename)); err != nil {
```
Comment (replay) [inline pkg/storage/tsdb/upload.go:68]:
```suggestion
	if err = objstore.UploadFile(ctx, logger, bkt, filepath.Join(blockDir, block.IndexFilename), path.Join(id.String(), block.IndexFilename)); err != nil {
```

File: pkg/storage/tsdb/upload.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+		return cleanUp(logger, bkt, id, errors.Wrap(err, "upload index"))
+	}
+
+	// Meta.json always need to be uploaded as a last item. This will allow to assume block directories without meta file to be pending uploads.
+	if err := bkt.Upload(ctx, path.Join(id.String(), block.MetaFilename), strings.NewReader(metaEncoded.String())); err != nil {
```
Comment (replay) [inline pkg/storage/tsdb/upload.go:73]:
```suggestion
	if err = bkt.Upload(ctx, path.Join(id.String(), block.MetaFilename), strings.NewReader(metaEncoded.String())); err != nil {
```

File: pkg/storage/tsdb/upload.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+		return strings.Compare(res[i].RelPath, res[j].RelPath) < 0
+	})
+	return res, err
+}
+
+func cleanUp(logger log.Logger, bkt objstore.Bucket, id ulid.ULID, err error) error {
+	// Cleanup the dir with an uncancelable context.
+	cleanErr := block.Delete(context.Background(), logger, bkt, id)
+	if cleanErr != nil {
+		return errors.Wrapf(err, "failed to clean block after upload issue. Partial block in system. Err: %s", err.Error())
```
Comment (replay) [inline pkg/storage/tsdb/upload.go:124]: doesn't that duplicate the error message?
It wraps the original error, but also adds the original error's message into the formatting string.

Comment (replay) [reply pkg/storage/tsdb/upload.go:124]: ah, i think it's supposed to wrap `cleanErr` instead of `err`, right?

Comment (replay) [reply pkg/storage/tsdb/upload.go:124]:
```suggestion
		return errors.Wrapf(cleanErr, "failed to clean block after upload issue. Partial block in system. Err: %s", err.Error())
```

File: pkg/storage/tsdb/upload.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		// and even though cleanUp will not see it yet, meta.json may appear in the bucket later.
+		// (Eg. S3 is known to behave this way when it returns 503 "SlowDown" error).
+		// If meta.json is not uploaded, this will produce partial blocks, but such blocks will be cleaned later.
+		return errors.Wrap(err, "upload meta file")
+	}
+
+	return nil
+}
+
+// Hopefully can be replaced with Thanos public function: https://github.com/thanos-io/thanos/pull/5400
```
Comment (replay) [inline pkg/storage/tsdb/upload.go:84]:
```suggestion
// gatherFileStats can hopefully be replaced with Thanos public function: https://github.com/thanos-io/thanos/pull/5400
```

File: pkg/storage/tsdb/upload.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+	res = append(res, mf)
+
+	metaFile, err := os.Stat(filepath.Join(blockDir, block.MetaFilename))
+	if err != nil {
+		return nil, errors.Wrapf(err, "stat %v", filepath.Join(blockDir, block.MetaFilename))
+	}
+	res = append(res, metadata.File{RelPath: metaFile.Name()})
+
+	sort.Slice(res, func(i, j int) bool {
+		return strings.Compare(res[i].RelPath, res[j].RelPath) < 0
```
Comment (replay) [inline pkg/storage/tsdb/upload.go:115]: The comment on the function `strings.Compare` says:

```
Compare is included only for symmetry with package bytes. It is usually clearer and always faster to use the built-in string comparison operators ==, <, >, and so on.
```

Comment (replay) [reply pkg/storage/tsdb/upload.go:115]:
```suggestion
		return res[i].RelPath < res[j].RelPath
```

**Review by pracucci — APPROVED** [review]
LGTM (modulo a couple of comments)

**Comment by pstibrany** [comment]
> > This PR modifies Shipper to avoid doing local copy of a block (using hardlinks) before uploading the block. Shipper was also updating its own copy of meta.json file. This has been replaced with
> 
> This sentence seems unfinished

Sorry about that, I fixed it.

File: pkg/ingester/shipper.go [hunk 11]
```diff
@@ -185,42 +187,22 @@ func (s *Shipper) Sync(ctx context.Context) (uploaded int, err error) {
...
 
-// sync uploads the block if not exists in remote storage.
-// TODO(khyatisoneji): Double check if block does not have deletion-mark.json for some reason, otherwise log it or return error.
+// update local meta.json file with Thanos-specific details, and upload the block
+// to blocks storage.
 func (s *Shipper) upload(ctx context.Context, meta *metadata.Meta) error {
 	level.Info(s.logger).Log("msg", "upload new block", "id", meta.ULID)
 
-	// We hard-link the files into a temporary upload directory so we are not affected
-	// by other operations happening against the TSDB directory.
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ingester/shipper.go:194]: This component comes from Thanos. Thanos sidecar runs next to Prometheus, and it is Prometheus itself managing the block. Prometheus can decide to delete the block for example. Shipper also updates meta.json file, and Prometheus could actually notice missing meta.json file (shipper deletes it first), if shipper worked on original file and not a copy.

In Mimir, block deletion is controlled by Mimir, and only done after block has been uploaded. In this PR we also don't modify meta.json file stored on disk.

File: pkg/ingester/shipper.go (hunk 3 above)
Comment (pstibrany) [reply pkg/ingester/shipper.go:197]: I meant to remove that, but left "and write" there by accident.

File: pkg/storage/tsdb/upload.go (hunk 4 above)
Comment (pstibrany) [reply pkg/storage/tsdb/upload.go:60]: I prefer version with reduced scope of new `err`. Not sure I have a good reason though.

File: pkg/storage/tsdb/upload.go (hunk 10 above)
Comment (pstibrany) [reply pkg/storage/tsdb/upload.go:115]: You catch everything! :) You should go through Mimir codebase and fix all lurking bugs there. (This was copied from Thanos again)

Comment (pstibrany) [reply pkg/storage/tsdb/upload.go:115]: Although this isn't strictly a bug

File: pkg/storage/tsdb/upload.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+// UploadBlock is copy of block.Upload with following modifications:
+//
+// - If meta parameter is supplied (not nil), then local meta.json file is ignored, and uploaded meta.json file reflects meta parameter.
+//
+// - Meta struct is updated with gatherFileStats
+//
+// - external labels are not checked for
+func UploadBlock(ctx context.Context, logger log.Logger, bkt objstore.Bucket, blockDir string, meta *metadata.Meta) error {
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/storage/tsdb/upload.go:33]: I've added unit test in the latest update. Please take a look. (It's copied from Thanos, modified to test updated functionality – esp. upload with in-memory meta.json file)

**Review by pracucci — APPROVED** [review]
Great job, LGTM!

**Review by replay — APPROVED** [review]
Looks good to me, thanks for addressing my feedback!

**Pull request merged**